src/scf/newton.jl

```diff
@@ -0,0 +1,317 @@
+# Newton's algorithm to solve SCF equations
+#
+# Newton algorithm consist of iterating over density matrices like
+#       P <- P - δP, with additional renormalization
+# where δP solves ([1])
+#       (Ω+K)δP = [P, [P, H(P)]]
+# where (Ω+K) if the constrained Hessian of the energy. It is
+# defined as a super operator on the space tangent to the constraints
+# manifold at point P_∞, the solution of our problem.
+#   - K is the unconstrained Hessian of the energy on the manifold;
+#   - Ω represents the influence of the curvature of the manifold :
+#         ΩδP = -[P_∞, [H(P_∞), δP]].
+#     In practice, we dont have access to P_∞ so we just use the current P.
+#     Another way to see Ω is as the four-point independent-particle
+#     susceptibility. Indeed, if we define the following extension and
+#     retraction operators
+#       - E : builds a density matrix Eρ given a density ρ via Eρ(r,r') = δ(r,r')ρ(r),
+#       - R : builds a density ρ given a density matrix O via ρ(r) = O(r,r),
+#     then we have the relation χ0₂ = -R(Ω₄^-1)E, where χ0₂ is the two_point
+#     independent-particle susceptibility (it returns δρ from a given δV) and
+#     Ω₄ is the four-point independent-particle susceptibility.
+#
+# For further details :
+# [1] Eric Cancès, Gaspard Kemlin, Antoine Levitt. Convergence analysis of
+#     direct minimization and self-consistent iterations. SIAM Journal of Matrix
+#     Analysis and Applications, 42(1):243–274 (2021)
+#
+# We presented the algorithm in the density matrix framework, but in practice,
+# we implement it with orbitals.
+# In this framework, an element on the tangent space at ψ can be written as
+#       δP = Σ |ψi><δψi| + hc
+# where the δψi are of size Nb and are all orthogonal to the ψj, 1 <= j <= N.
+# Therefore we store them in the same kind of array than ψ, with
+# δψ[ik][:, i] = δψi for each k-point.
+# In this framework :
+#   - computing Ωδψ can be done analytically with the formula
+#         Ωδψi = H*δψi - Σj λji δψj
+#     where λij = <ψi|H|ψj>. This is one way among others to extend the
+#     definition of ΩδP = -[P_∞, [H(P_∞), δP]] to any point P on the manifold --
+#     we chose this one because it makes Ω self-adjoint for the Frobenius scalar
+#     product;
+#   - computing Kδψ can be done by constructing the δρ associated with δψ, then
+#     using the exchange-correlation kernel to compute the associated δV, then
+#     acting with δV on the ψ and projecting on the tangent space.
+
+using LinearMaps
+using IterativeSolvers
+
+#  Compute the gradient of the energy, projected on the space tangent to ψ, that
+#  is to say H(ψ)*ψ - ψ*λ where λ is the set of Rayleigh coefficients associated
+#  to the ψ.
+function compute_projected_gradient(basis::PlaneWaveBasis, ψ, occupation)
+    ρ = compute_density(basis, ψ, occupation)
+    _, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
+
+    [proj_tangent_kpt(H.blocks[ik] * ψ[ik], ψ[ik])
+     for ik = 1:length(basis.kpoints)]
```

Comment:
I prefer to write such contractions as
```suggestion
    [proj_tangent_kpt(H.blocks[ik] * ψk, ψk) for (ik, ψk) in enumerate(ψ)]
```